components/input/Input.js

```diff
@@ -0,0 +1,161 @@
+import React, { Component, createElement } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import { IconChevronUpSmallOutline, IconChevronDownSmallOutline, IconErrorSmallOutline } from '@teamleader/ui-icons';
+import InputMetaPropTypes from './InputMetaPropTypes';
+import Counter from '../counter';
+import theme from './theme.css';
+
+export default class Input extends Component {
+  static propTypes = {
+    bold: PropTypes.bool,
+    className: PropTypes.string,
+    counter: PropTypes.number,
+    disabled: PropTypes.bool,
+    icon: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
+    iconPlacement: PropTypes.oneOf(['left', 'right']),
+    id: PropTypes.string,
+    meta: InputMetaPropTypes,
+    onBlur: PropTypes.func,
+    onChange: PropTypes.func,
+    onFocus: PropTypes.func,
+    placeholder: PropTypes.string,
+    readOnly: PropTypes.bool,
+    size: PropTypes.oneOf(['small', 'medium', 'large']),
+    step: PropTypes.number,
+    type: PropTypes.string,
+    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  };
+
+  static defaultProps = {
+    iconPlacement: 'left',
+    disabled: false,
+    placeholder: '',
+    readOnly: false,
+    size: 'medium',
+    step: 1,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.isNumberInput = Boolean(props.type === 'number');
+    this.handleChange = this.onChange.bind(this);
+    this.handleIncreaseValue = this.increaseValue.bind(this);
+    this.handleDecreaseValue = this.decreaseValue.bind(this);
+
+    this.state = {
+      value: this.isNumberInput ? Number(props.value) || '' : props.value || '',
+    };
+  }
+
+  componentWillUpdate(props, state) {
+    if (state.value) {
+      this.props.onChange(state.value);
+    }
+  }
+
+  onChange(value) {
+    if (this.isNumberInput) {
+      this.setState({ value: Number(value.target.value) });
+    } else {
+      this.setState({ value: value.target.value });
+    }
+  }
+
+  increaseValue() {
+    if (this.state.value) {
+      return this.setState({ value: this.state.value + this.props.step });
+    }
+
+    return this.setState({ value: this.props.step });
+  }
+
+  decreaseValue() {
+    if (this.state.value) {
+      return this.setState({ value: this.state.value - this.props.step });
+    }
```

Comment:
As you know, React may batch multiple `setState()` calls into a single update for performance. Could you change this to the setState with function notation, just or my ease of mind?

```jsx
this.setState((prevState, props) => ({ value: prevState.value ? prevState.value - props.step : -props.step }));
```

(And of course everywhere else where comparing with previous state)

Reply:
If it eases your mind, sure!